# 01 - Data Understanding

## Online Retail Sales & Customer Analysis

This notebook focuses on understanding the structure, content, quality, and business meaning of the Online dataset before performing data cleaning and exploratory analysis.

## 1. Project Objective

The objective of this project is to analyze transaction data from an international online retailer to understand:

- sales and revenue performance
- product performance
- customer perchasing behaviour
- geograohical patterns
- customer value and segmentation

Before performing analysis, we first need to understand the dataset, its variables, its structure, and potential data quality issues.

In [2]:
import pandas as pd

# Load the dataset
df = pd.read_excel("../OnlineRetail.xlsx")

## 2. First Look at the Dataset

We begin by inspecting a few records to understand what on e row represents and what information is availables for each transaction.

In [3]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [4]:
df.shape

(541909, 8)

In [5]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='str')

## 3. Data Structure

We inspect the dataset's schema, including the number of observations, data types, and missing values.

In [11]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[us]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 33.1+ MB


In [7]:
df.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

### Initital observations

- The dataset contains 541,909 transactions.
- TThere are 8 columns.
- 'CustomerID' contains missing values.
- 'Description' contains missing values.
- 'InvoiceDate' is already stored as a datetime type.
- 'Quantity' is integer-valued.
- 'UnitePrice' is floating-point.

## 4. Descriptive Statistics

Descriptive statistics provide an initial view of the numerical variables and help identify unusual values or distributions that require further investigation.

In [14]:
df.describe()

,Quantity,InvoiceDate,UnitPrice,CustomerID
count,541909.000000,541909,541909.000000,406829.000000
mean,9.552250,2011-07-04 13:34:57.156386,4.611114,15287.690570
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,218.081158,NaN,96.759853,1713.600303


## 5. Understanding the Dataset's Business Entities

The dataset contains information at the transaction-linelevel.
We therefore examine thenumber of unique invoices, customers, products,andcountries

This helps distinguish between transaction records and higher-level business entities such as orders and customers.

In [16]:
print("Unique invoices:", df["InvoiceNo"].nunique())
print("Unique customers:", df["CustomerID"].nunique())
print("Unique products:", df["StockCode"].nunique())
print("Unique countries:", df["Country"].nunique())

Unique invoices: 25900
Unique customers: 4372
Unique products: 4070
Unique countries: 38


## 6. Geographic Coverage

We examine the countries represented in the dataset to determine
whether the retailer operates only in its domestic UK market or
serves an international customer base.

In [18]:
df["Country"].nunique()
df["Country"].unique()
df["Country"].value_counts()

Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


## 7. Initial Data Quality Investigation

### 7.1 Negative Quantities

The `Quantity` variable contains negative values. Since quantity would
normally represent the number of units purchased, these observations
require investigation before deciding how they should be treated.

In [19]:
df["Quantity"].describe()

count    541909.000000
mean          9.552250
std         218.081158
min      -80995.000000
25%           1.000000
50%           3.000000
75%          10.000000
max       80995.000000
Name: Quantity, dtype: float64

In [29]:
print((df["Quantity"] < 0).sum())

10624


In [22]:
df[df["Quantity"] < 0].count()

InvoiceNo      10624
StockCode      10624
Description     9762
Quantity       10624
InvoiceDate    10624
UnitPrice      10624
CustomerID      8905
Country        10624
dtype: int64

In [20]:
df[df["Quantity"] < 0].head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom


In [23]:
df["InvoiceNo"].astype(str).str.startswith("C").value_counts()

InvoiceNo
False    532621
True       9288
Name: count, dtype: int64

In [24]:
df.loc[
    df["InvoiceNo"].astype(str).str.startswith("C"),
    "Quantity"
].describe()

count     9288.000000
mean       -29.885228
std       1145.786965
min     -80995.000000
25%         -6.000000
50%         -2.000000
75%         -1.000000
max         -1.000000
Name: Quantity, dtype: float64

### 7.2 Unit Price

The `UnitPrice` variable contains zero, negative, and extreme values.
These observations need to be investigated to determine whether they
represent valid business transactions, special cases, or data quality
issues.

In [25]:
df["UnitPrice"].describe()

count    541909.000000
mean          4.611114
std          96.759853
min      -11062.060000
25%           1.250000
50%           2.080000
75%           4.130000
max       38970.000000
Name: UnitPrice, dtype: float64

In [26]:
print("Zero prices:", (df["UnitPrice"] == 0).sum())
print("Negative prices:", (df["UnitPrice"] < 0).sum())
print("Prices above 100:", (df["UnitPrice"] > 100).sum())

Zero prices: 2515
Negative prices: 2
Prices above 100: 1036


In [ ]:
df[df["UnitPrice"] > 100][
    [
        "InvoiceNo",
        "StockCode",
        "Description",
        "Quantity",
        "UnitPrice",
        "CustomerID",
        "Country"
    ]
].head(20)

## 8. Initial Findings

Based on the initial investigation, we identified several areas
requiring further investigation before data cleaning:

- 
- 
- 

## 9. Open Questions Before Data Cleaning

Before modifying the dataset, we need to determine:

1. What do negative quantities represent?
2. What do zero and negative unit prices represent?
3. How should cancellation transactions be treated?
4. How should missing CustomerID values affect customer-level analysis?
5. Are there duplicate transaction records?
6. Are extreme values legitimate business transactions?